### ISSUE TITLE
OAuth 2.0 Device Authorization Grant (RFC 8628)

### ISSUE THREAD

**Issue by sjoerdsimons** [kanidm/kanidm#1523]
**Is your feature request related to a problem? Please describe.**

Feature request

**Describe the solution you'd like**

 OAuth 2.0 Device Authorization Grant provide a nice way for e.g. command line apps and remote clients to trigger authorization against an oidc/oath service without having to require a local webbrowser.

**Comment by yaleman** [comment]
I support this, it'd allow for things like passkey-backed SSH/CLI usage without having to build it into the client applications.

**Comment by Firstyear** [comment]
This has certainly been on my radar for some time, I just hadn't made it a priority yet. I'm currently focused on refresh tokens and then some idm sync stuff, maybe once that's a bit further I'll come back to this.

**Comment by Firstyear** [comment]
https://www.rfc-editor.org/rfc/rfc8628

**Comment by Firstyear** [comment]
@yaleman Can you explain more about the ssh/cli use case for me? I'm curious about it.

**Comment by yaleman** [comment]
> @yaleman Can you explain more about the ssh/cli use case for me? I'm curious about it.

Duo's module would be the one to look at for a comparison - https://duo.com/docs/duounix#test-your-setup

It prompts for auth as one of the factors and in our case we'd be starting the device auth flow and prompting the user to open a url to complete the auth.

**Comment by Firstyear** [comment]
So it prompts for unix password I guess then does the device flow for the second/extra factors? 

I think given how device auth works, we'll have to "preauth" the flow somehow in this case with the unix password. There are some challenges to make device auth flows work nicely in a distributed system so I need to think more about this. But I think it certainly would have uses.

**Comment by yaleman** [comment]
yeah, there's a bunch of ways you could implement it, for sure - and PAM would allow you to entirely replace the auth flow with our module if you really wanted to. IIRC the default they recommend is still a UNIX password with them for second factor, but was designed before "passwordless" flows were common.

**Comment by Firstyear** [comment]
So I'm guessing the value here is for non-kanidm ssh clients/auth? Because in Kanidm we'll be able to do passwordless flows with attested ssh keys in the future so it's not super important on that front.

**Comment by yaleman** [comment]
At some point you're going to have to authenticate, for things like sudo I guess?

**Comment by yaleman** [comment]
> in Kanidm we'll be able to do passwordless flows with attested ssh keys in the future so it's not super important on that front.

I might need some info about what *this* looks like before I can really respond though 😄

**Comment by ieugen** [comment]
Would love to see this implemented. 
I would like to implement Device Authorization Grant to handle authentication in a CLI used for devops purposes. 

Since devs will have accounts in KaniDM I would like to register an Oauth2 app in KaniDM that supports Device Grant. 
Users will ssh into servers and use the CLI.
The cli will display Device Grant URL + code, users will authenticate using browser and have temporary access to call API and do what they need to do.

**Comment by johnnypea** [comment]
Do you have a date when this can be implemented? Thank you.

**Comment by Firstyear** [comment]
No, there are other priorities right now.

**Comment by helmut72** [comment]
Is there some kind of priority/roadmap?

**Comment by Blacks-Army** [comment]
Hey, are there any news?

**Comment by yaleman** [comment]
References is now possible, so this can be worked on now, I'll be working on it as I have time.

**Comment by SigmaSquadron** [comment]
Hey folks; just to make sure: is this issue also tracking the discovery of `DeviceAuthURL` from the OpenID `.well-known` endpoint, or will there be some other interface for the device flow?

**Comment by Firstyear** [comment]
Once the feature is added, we'll add it to the discovery URL's too.

**Comment by djds** [comment]
Not to add noise here, but a recent use case that I encountered was being able to set up [OIDC for Incus](https://linuxcontainers.org/incus/docs/main/authentication/#authentication-openid), which requires the device authorization grant. Anyway, thanks for the great software!
